pkg/util/k8scmcache.go

```diff
@@ -0,0 +1,208 @@
+        }
+        return namespace
+}
+
+func NewK8sClient() (*k8s.Clientset) {
+        var cfg *rest.Config
+        var err error
+        cPath := os.Getenv("KUBERNETES_CONFIG_PATH")
+        if cPath != "" {
+                cfg, err = clientcmd.BuildConfigFromFlags("", cPath)
+                if err != nil {
+                        glog.Errorf("Failed to get cluster config with error: %v\n", err)
+                        os.Exit(1)
+                }
+        } else {
+                cfg, err = rest.InClusterConfig()
+                if err != nil {
+                        glog.Errorf("Failed to get cluster config with error: %v\n", err)
+                        os.Exit(1)
+                }
+        }
+        client, err := k8s.NewForConfig(cfg)
+        if err != nil {
+                glog.Errorf("Failed to create client with error: %v\n", err)
+                os.Exit(1)
+        }
+        return client
+}
+
+func (k8scm *K8sCMCache) createMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+        cm := &v1.ConfigMap{}
+        cm, err := k8scm.getMetadataCM(resourceID)
+        if err != nil {
+                glog.Infof("k8s-cm-cache: couldn't get configmap %s with error: %v", resourceID, err)
+                glog.Infof("k8s-cm-cache: creating configmap...")
+                cm = &v1.ConfigMap {
+                        ObjectMeta: metav1.ObjectMeta {
+                                Name: resourceID,
+                                Namespace: k8scm.Namespace,
+                                Labels: map[string]string{
+                                        csiMetadataLabelAttr: cmLabel,
+                                },
+                        },
+                        Data: map[string]string{},
+                }
+                cm, err = k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Create(cm)
+                if err != nil {
+                        glog.Errorf("k8s-cm-cache: couldn't create configmap %s with error: %v\n", resourceID, err)
+                        return nil, errors.Wrap(err, "k8s-cm-cache: create configmap error")
+                }
+        }
+        glog.Infof("k8s-cm-cache: configmap %s successfully created\n", resourceID)
+        return cm, nil
+}
+
+func (k8scm *K8sCMCache) getMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+	k8sCMCacheMtx.Lock()
+	defer k8sCMCacheMtx.Unlock()
+
+        cm, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Get(resourceID, metav1.GetOptions{})
+        if err != nil {
+                return nil, err
+        }
+        return cm, nil
+}
+
+
+func (k8scm *K8sCMCache) ForAll(pattern string, destObj interface{}, f ForAllFunc) error {
+	listOpts := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", csiMetadataLabelAttr, cmLabel)}
+	cms, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).List(listOpts)
+        if err != nil {
+                glog.Infof("k8s-cm-cache: failed to list metadata configmaps")
```

Comment:
To keep things consistent, don't log here but rather wrap the error message. The same goes for all other occurrences of logging&returning errors. We can log errors outside (e.g. in the `controllerserver.go`)